### PR TITLE
improve date manipulation performance

### PR DESCRIPTION
**Pull request by bs85**
After debugging performance issues with a very heavily loaded calendar (~300 events per day across ~15 resources, 3 months of events passed in the `events` variable), I realized a couple functions were responsible for over 90% of the processing time.

This version is an order of magnitude faster for my use-case, going from ~900ms to < 70ms to render a heavily loaded week-view.

The main culprit I found was `date-arithmetic`, especially the `startOf` function, that while written very elegantly has absolutely horrible performance as it mutates the date many times.

While debugging it, I also realized `eventLevels.inRange` implementation was very costly, and could be sped up 10-20x by changing the input params a bit (passing the real range we want to filter and just using `<` and `>`).

I also added a autoDetectAllDayEvents prop to the calendar, that is true by default. Changing it to false will stop analysing every event time to see if both start and end are 00:00:00.

This works theoretically and I've tested it with my dataset, but there might be corner cases I'm not thinking of (zero time events combined with something else, daylight savivng, etc?).

Thanks in advance for having an in-depth look!

### CODE REVIEW

File: src/utils/eventLevels.js [hunk 1]
```diff
@@ -55,17 +55,14 @@ export function eventLevels(rowSegments, limit = Infinity) {
...
-export function inRange(e, start, end, accessors) {
-  let eStart = dates.startOf(accessors.start(e), 'day')
-  let eEnd = accessors.end(e)
+export function inRange(e, lowerBound, upperBound, accessors) {
+  const start = accessors.start(e)
+  const end = accessors.end(e)
 
-  let startsBeforeEnd = dates.lte(eStart, end, 'day')
-  // when the event is zero duration we need to handle a bit differently
-  let endsAfterStart = !dates.eq(eStart, eEnd, 'minutes')
```
Comment (jquense) [inline src/utils/eventLevels.js]: this line was preventing a ton of edge case bugs, i'm hesitant to just remove it without understanding how the new logic is equivalent.

Comment (jquense) [reply src/utils/eventLevels.js]: also probably fine to do a straight getTime() comparison vs minutes

File: src/utils/dates.js [hunk 2]
```diff
@@ -145,12 +153,18 @@ let dates = {
     return dates.startOf(new Date(), 'day')
   },
 
-  yesterday() {
-    return dates.add(dates.startOf(new Date(), 'day'), -1, 'day')
+  tomorrow(date) {
```
Comment (jquense) [inline src/utils/dates.js]: how about `dayAfter` for a name. tomorrow to me always means the day after today :P

File: src/utils/dates.js [hunk 3]
```diff
@@ -145,12 +153,18 @@ let dates = {
...
+    const tomorrow = new Date(date.valueOf())
+    tomorrow.setDate(tomorrow.getDate() + 1)
+
+    return tomorrow
   },
 
-  tomorrow() {
-    return dates.add(dates.startOf(new Date(), 'day'), 1, 'day')
+  yesterday(date) {
```
Comment (jquense) [inline src/utils/dates.js]: `dayBefore`...are there any existing usages of this that would have broke?

File: src/TimeGrid.js [hunk 4]
```diff
@@ -203,15 +205,19 @@ export default class TimeGrid extends Component {
     let allDayEvents = [],
       rangeEvents = []
 
+    const upperBound = new Date(end.valueOf())
```
Comment (jquense) [inline src/TimeGrid.js]: `dates.tomorrow`?

File: src/utils/dates.js [hunk 5]
```diff
@@ -145,12 +153,18 @@ let dates = {
     return dates.startOf(new Date(), 'day')
   },
 
-  yesterday() {
-    return dates.add(dates.startOf(new Date(), 'day'), -1, 'day')
+  tomorrow(date) {
+    const tomorrow = new Date(date.valueOf())
```
Comment (jquense) [inline src/utils/dates.js]: can we use `getTime()` or `+date` instead of `valueOf`

File: src/utils/eventLevels.js [hunk 6]
```diff
@@ -55,17 +55,14 @@ export function eventLevels(rowSegments, limit = Infinity) {
   return { levels, extra }
 }
 
-export function inRange(e, start, end, accessors) {
-  let eStart = dates.startOf(accessors.start(e), 'day')
-  let eEnd = accessors.end(e)
+export function inRange(e, lowerBound, upperBound, accessors) {
```
Comment (jquense) [inline src/utils/eventLevels.js]: can you add a comment noting the performance concerns of this method as an explanation for why it does use `date-arithmetic`, so someone doesn't accidentally remove the code later :P

**Comment by jquense** [comment]
this is awesome thanks 👍

**Comment by bs85** [comment]
Thanks a lot for the comments, will cleanup style and add comments, and deal with the failing tests ASAP

**Comment by jquense** [comment]
any chance to clean this up @bs85 ?

**Comment by bs85** [comment]
Not yet sorry. Do you want to close it and re-open when I can get to it? Might not be until a couple more weeks.

**Comment by jquense** [comment]
I'm happy to leave it, I just didn't want to close if your planning on getting to it. no rush of course

File: src/utils/eventLevels.js (hunk 1 above)
Comment (bs85) [reply src/utils/eventLevels.js]: Changed my implementation slightly. upperBound is now exclusive, and in case the event start is equal to upperBound, we check if the event has a zero duration, to ensure that the function always returns true if the range and the event have the exact same bounds.

Could you say more about those edge cases? This function is only called in 3 locations (in Month.js, Agenda.js and TimeGrid.js). Other parts of the code call the `inRange` method of `utils/date`.

File: src/utils/dates.js (hunk 2 above)
Comment (bs85) [reply src/utils/dates.js]: Done

File: src/TimeGrid.js (hunk 4 above)
Comment (bs85) [reply src/TimeGrid.js]: Done

File: src/utils/eventLevels.js (hunk 6 above)
Comment (bs85) [reply src/utils/eventLevels.js]: Commented the inRange method

File: src/utils/dates.js (hunk 5 above)
Comment (bs85) [reply src/utils/dates.js]: using `getTime` everywhere now

File: src/utils/dates.js (hunk 3 above)
Comment (bs85) [reply src/utils/dates.js]: It wasn't used anywhere, the part that used it was probably changed

**Comment by bs85** [comment]
Found some time to clean it up!

As the `inRange` method's behavior changed, I also rewrote the tests based on the assumptions of the new version.

I think `eventLevels.inRange` could reasonably be considered a private method, but in practice some projects might be importing it and using it, especially in custom components.

I overwrote it in this version, but if you think it is too dangerous, I can also create a new `fastInRange` method, and deprecate `inRange` for now but keep it.

File: test/utils/eventLevels.test.js [hunk 7]
```diff
@@ -261,107 +261,71 @@ describe('inRange', () => {
         expect(inRange(event, rangeStart, rangeEnd, accessors)).toBe(result)
       })
     }
-    const weekOfThe5th = [d(5), d(11)]
-    const weekOfThe12th = [d(12), d(18)]
-    ;[
+
+    const sample = [
```
Comment (bs85) [inline test/utils/eventLevels.test.js]: prettier was prepending a semi-colon here, and eslint was complaining. Assigning to `sample` as a workaround

File: src/utils/eventLevels.js (hunk 1 above)
Comment (jquense) [reply src/utils/eventLevels.js]: > Could you say more about those edge cases? 

There are a few tests that cover them, but it mostly comes down to events that span weeks or sit on the leading/trailing border of a week. If you don't account for zero duration events specifically they can get dropped from the week or show up at the start and end. If tests are passing tho, that should indicate the case is covered (should)
